harness/determined/core/_profiler.py

```diff
@@ -0,0 +1,412 @@
+import abc
+import logging
+import queue
+import threading
+import time
+from datetime import datetime, timezone
+from typing import Any, Dict, List
+
+import psutil
+
+from determined import core
+from determined.common import api, constants
+from determined.common.api import bindings
+
+try:
+    import pynvml
+except ImportError:
+    pynvml = None
+
+logger = logging.getLogger("determined.core")
+
+
+class ProfilerContext:
+    """
+    ``ProfilerContext`` gives access to the system profiling feature within Determined.
```

Comment:
In an ideal world where we had all the time we could ever want into writing docs, I'd like this to class docstring to give a usage example. I think I'll be able to figure it out by looking at `_tf_keras_trial.py`, but I don't _want_ to have to look at examples to understand the class.

This is ostensibly part of the public interface, so I think writing a usage example directly into the docstring would be nice. But I'd talk this out if you think it's bad or not worth it.

Reply:
Thanks. This gets me part of the way there. I still don't know how I'm supposed to _use_ the `ProfilerContext` in my code.

I can't tell if I'm being obtuse or if documentation is missing. Chat?

Reply:
I think this is really good.